Tests/AppTests/PipelineTests.swift

```diff
@@ -0,0 +1,208 @@
+            XCTAssertEqual(packages.map(\.url), ["1", "3", "4"].gh)
+            XCTAssertEqual(packages.map(\.status), [.ok, .ok, .none])
+            XCTAssertEqual(packages.map(\.processingStage), [.analysis, .analysis, .reconciliation])
+        }
+
+        // MUT - ingest again
+        try ingest(client: app.client, database: app.db, limit: 10).wait()
+
+        do {  // validate - only new package moves to .ingestion stage
+            let packages = try Package.query(on: app.db).sort(\.$url).all().wait()
+            XCTAssertEqual(packages.map(\.url), ["1", "3", "4"].gh)
+            XCTAssertEqual(packages.map(\.status), [.ok, .ok, .ok])
+            XCTAssertEqual(packages.map(\.processingStage), [.analysis, .analysis, .ingestion])
+        }
+
+        // MUT - analyze again
+        let lastAnalysis = Current.date()
+        try analyze(application: app, limit: 10).wait()
+
+        do {  // validate - only new package moves to .ingestion stage
+            let packages = try Package.query(on: app.db).sort(\.$url).all().wait()
+            XCTAssertEqual(packages.map(\.url), ["1", "3", "4"].gh)
+            XCTAssertEqual(packages.map(\.status), [.ok, .ok, .ok])
+            XCTAssertEqual(packages.map(\.processingStage), [.analysis, .analysis, .analysis])
+            XCTAssertEqual(packages.map { $0.updatedAt! > lastAnalysis }, [false, false, true])
+        }
+
+        // fast forward our clock by the deadtime interval
+        Current.date = { Date().addingTimeInterval(Constants.reingestionDeadtime) }
+
+        // MUT - ingest yet again
+        try ingest(client: app.client, database: app.db, limit: 10).wait()
+
+        do {  // validate - now all three packages should have been updated
+            let packages = try Package.query(on: app.db).sort(\.$url).all().wait()
+            XCTAssertEqual(packages.map(\.url), ["1", "3", "4"].gh)
+            XCTAssertEqual(packages.map(\.status), [.ok, .ok, .ok])
+            XCTAssertEqual(packages.map(\.processingStage), [.ingestion, .ingestion, .ingestion])
+        }
+
+        // MUT - re-run analysis to complete the sequence
+        try analyze(application: app, limit: 10).wait()
+
+        do {  // validate - only new package moves to .ingestion stage
+            let packages = try Package.query(on: app.db).sort(\.$url).all().wait()
+            XCTAssertEqual(packages.map(\.url), ["1", "3", "4"].gh)
+            XCTAssertEqual(packages.map(\.status), [.ok, .ok, .ok])
+            XCTAssertEqual(packages.map(\.processingStage), [.analysis, .analysis, .analysis])
+        }
+
+        // at this point we've ensured that retriggering ingestion after the deadtime will
+        // refresh analysis as expected
+    }
+
+}
+
+
+private extension String {
+    var gh: String { "https://github.com/foo/\(self)" }
+}
+
+private extension Array where Element == String {
+    var gh: Self { map(\.gh) }
+}
```

Comment:
Just two little helpers to make the tests more readable by turning "1" into a url that can be processed without cluttering everything up with lots of `https://github.com/foo/1` ceremony.